### PR TITLE
Add DataSourceRef field to PVC spec

### LINKED ISSUE #103275: Update volume populators for 1.22
#### What would you like to be added:

Implement up updated version of the volume populators KEP, which modifies the PVC API.

#### Why is this needed:

KEP 1495 was updated for the 1.22 release with a new design:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators

### PR DESCRIPTION
**Pull request by bswartz**
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

Modify the behavior of the AnyVolumeDataSource alpha feature gate to enable
a new field, DataSourceRef, rather than modifying the behavior of the
existing DataSource field. This allows addition Volume Populators in a way
that doesn't risk breaking backwards compatibility, although it will
result in eventually deprecating the DataSource field.

#### Which issue(s) this PR fixes:
Fixes #103275

#### Special notes for your reviewer:

A lot of test cases have been added, but I think more could be added. I'm going to keep working on those which the change itself is reviewed.

#### Does this PR introduce a user-facing change?
```release-note
Add DataSourceRef alpha field to PVC spec, which allows contents other than PVCs and VolumeSnapshots to be data sources.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators
```

### CODE REVIEW

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -450,13 +450,21 @@ type PersistentVolumeClaimSpec struct {
...
+	// If the AnyVolumeDataSource feature gate is enabled, this field will always have
+	// the same contents as the DataSourceRef field.
 	// +optional
 	DataSource *TypedLocalObjectReference
+	// Specifies the object from which to populate the volume with data, if a non-empty
+	// volume is desired. This may be any local object from a non-empty API group (non
+	// core object) or a PersistentVolumeClaim object.
+	// When this field is specified, volume binding will only succeed if the type of
+	// the specified object matches some installed volume populator or dynamic
+	// provisioner.
```
Comment (thockin) [inline pkg/apis/core/types.go:464]: I think you need to explain the special relationship to the DataSource field

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -492,13 +492,21 @@ type PersistentVolumeClaimSpec struct {
...
 	// it will create a new volume based on the contents of the specified data source.
+	// If the AnyVolumeDataSource feature gate is enabled, this field will always have
+	// the same contents as the DataSourceRef field.
 	// +optional
 	DataSource *TypedLocalObjectReference `json:"dataSource,omitempty" protobuf:"bytes,7,opt,name=dataSource"`
+	// Specifies the object from which to populate the volume with data, if a non-empty
+	// volume is desired. This may be any local object from a non-empty API group (non
+	// core object) or a PersistentVolumeClaim object.
+	// When this field is specified, volume binding will only succeed if the type of
+	// the specified object matches some installed volume populator or dynamic
```
Comment (thockin) [inline staging/src/k8s.io/api/core/v1/types.go:505]: Same as above - explain the multi-field realtionship, pl3ase

File: pkg/registry/core/persistentvolumeclaim/storage/storage.go [hunk 3]
```diff
@@ -73,6 +77,43 @@ func (r *REST) ShortNames() []string {
 	return []string{"pvc"}
 }
 
+// defaultOnRead sets interlinked fields that were not previously set on read.
+// We can't do this in the normal defaulting path because that same logic
+// applies on Get, Create, and Update, but we need to distinguish between them.
+//
+// This will be called on both PersistentVolumeClaim and PersistentVolumeClaimList types.
+func (r *REST) defaultOnRead(obj runtime.Object) {
```
Comment (thockin) [inline pkg/registry/core/persistentvolumeclaim/storage/storage.go:85]: if-you-like-it-then-you-should-have-put-a-test-on-it

File: pkg/api/persistentvolumeclaim/util.go [hunk 4]
```diff
@@ -67,3 +68,20 @@ func dataSourceIsEnabled(pvcSpec *core.PersistentVolumeClaimSpec) bool {
...
+
+// NormalizeDataSources ensures that DataSource and DataSourceRef have the same contents
+// as long as both are not explicitly set.
+// This should be used by creates/gets of PVCs, but not updates
+func NormalizeDataSources(pvcSpec *core.PersistentVolumeClaimSpec) {
+	// Don't enable this behavior if the feature gate is not on
+	if !utilfeature.DefaultFeatureGate.Enabled(features.AnyVolumeDataSource) {
+		return
+	}
+	if pvcSpec.DataSource != nil && pvcSpec.DataSourceRef == nil {
```
Comment (thockin) [inline pkg/api/persistentvolumeclaim/util.go:80]: This isn't clearly what the KEP calls for.  It may actually WORK but a reader of this will not comprehend it.

Where is the code that clears DataSource when it's not a PVC or Snapshot?  Should that be here?  Otherwise what happens if I send `{ DataSource = Invalid, DataSourceRef = nil }`?  Invalid will get copied to DSRef, then (somewhere) DS will be cleared, then DS and DSRef are not the same.  This converts case 5 (in KEP) into case 6.

Basically, the compat logic is so precise and subtle that I think you should centralize it and comment the heck out of it.  Take pity on the next poor sucker who deals with this.

Comment (bswartz) [reply pkg/api/persistentvolumeclaim/util.go:80]: So the case you mention is covered by the fact that DropDisabledFields() is called before NormalizeDataSources() in PrepareForCreate() in stragtegy.go. I agree it's not easy to follow, but there are substantial tests already for the existing code, so there's a strong reason not to make a ton of changes to the existing code, and to make these new changes in a mostly additive way. I can try to address the problem with lots and lots of comments.

File: pkg/registry/core/persistentvolumeclaim/storage/storage.go (hunk 3 above)
Comment (bswartz) [reply pkg/registry/core/persistentvolumeclaim/storage/storage.go:85]: Good point, I tested the underlying function but this layer deserves a few unit tests. E2E tests will help as well.

File: pkg/api/persistentvolumeclaim/util.go (hunk 4 above)
Comment (thockin) [reply pkg/api/persistentvolumeclaim/util.go:80]: By convention, the DropDisabledFields pattern holds feature-gated logic.  No feature-gates == no logic.  I'd STRONGLY encourage you to gather all of the "permanent" logic together.  Having to jump between functions to convince myself this works is unpleasant at best (and the tests are all split) and very brittle.  I can EASILY see logic around here getting "fixed" wrongly because it's not at all clear how it is supposed to work or why.

**Review by thockin — COMMENTED** [review]
I'll be online off and on today and tomorrow.  If you want to make an argument for "the tests work, let's merge it and then fix this" I am open to it, but I really REALLY think it should per made more permanent, since it is now part of the defined semantics.

**Comment by bswartz** [comment]
@thockin Thanks I've been working on refactoring this code and adding more unit tests today. I'll send up another version trying to incorporate all your suggestions. The one tricky one will be the DropDisabledFields() implementation -- I understand it's intended to hold the feature-gated logic, but the existing implementation already doesn't do that. Given the ugly reality of this I will try to hold to the spirit of that goal and you can tell me what you think. I'll ping you when I push another commit this afternoon.

**Comment by bswartz** [comment]
That last commit represents the changes to DropDisabledFields() and more tests. I still haven't reworded the comments/docs for the new core field. That will come next.

**Comment by bswartz** [comment]
The most recent commit also updates the doc comments, and should cover all of the feedback so far. Please let me know if the rewrite of DropDisabledFields() is clear enough and follows the spirit of the feature-gate rule. The original logic was flawed so this new version is not only more readable but also more functional.

File: pkg/api/persistentvolumeclaim/util.go [hunk 5]
```diff
@@ -39,26 +65,21 @@ func dataSourceInUse(oldPVCSpec *core.PersistentVolumeClaimSpec) bool {
...
 		return false
 	}
-	if oldPVCSpec.DataSource != nil {
+	if oldPVCSpec.DataSource != nil || oldPVCSpec.DataSourceRef != nil {
 		return true
 	}
 	return false
 }
 
 func dataSourceIsEnabled(pvcSpec *core.PersistentVolumeClaimSpec) bool {
```
Comment (thockin) [inline pkg/api/persistentvolumeclaim/util.go:74]: IsEnabled is now a total misnomer.

**Review by thockin — COMMENTED** [review]
This actually pushed me farther down the "move it" road.  IIUC, updates of PVC can't change the data source, so having this logic in DropDisabled, which has an "old" and "new" param is extra confusing.  Putting it in Normalize, which is only called on Create will make it that much more obvious what the heck is going on.

See, there's a story that you need to tell here.  3, 4, 5 years from now, someone will be staring at this and asking "WTF?!", running `git blame` and muttering about you.  Very possibly that someone will be you. 

I think you should move it, and tell the whole story in Normalize.  You are swimming upstream.

If you need to merge this and followup, I could buy that, but I REALLY think we will regret not doing it right.

File: pkg/api/persistentvolumeclaim/util.go (hunk 5 above)
Comment (bswartz) [reply pkg/api/persistentvolumeclaim/util.go:74]: DropDisabled is called by Create and Update, and Normalize is called by Create and Get/List. I'm not opposed to more refactoring, but I think we'll end up with a 3rd function and the unit tests will need a complete rewrite as a result.

I also agree that the existing method names suck, but there's a tension between wanting to clean up bad code and wanting to avoid fixing stuff that's not broken.

**Comment by bswartz** [comment]
Okay I have split most of the code in DropDisabledFields into another function, so Drop just handles the feature gate dropping. The new function is called EnforceDataSourceBackwardsCompatibility (long, but hopefully not confusing). I don't think it's possible to combine this new function with NormalizeDataSources, because Normalize is also used on the Get/List code path.

Lots of unit tests needed updating. And also I realized that updates needed some work so I fixed that code path too. In particular we have to be careful not to break any updates which would have previously worked, given that the validation code prevents mutations of PVC specs, and we're playing a lot of games with automatically modifying these fields.

I still want to add a raft of E2E tests that cover the combination of all the changes, to try to smoke out any corner cases, especially around combinations of updates and changing of the feature gate from off to on or on to off.

File: pkg/api/persistentvolumeclaim/util.go (hunk 5 above)
Comment (thockin) [reply pkg/api/persistentvolumeclaim/util.go]: It *is* broken, it's just broken in human-space not compiler-space :)

File: pkg/api/persistentvolumeclaim/util.go [hunk 6]
```diff
@@ -29,34 +29,25 @@ const (
...
-	// gate is disabled or the dataSourceRef field is not filled in, then
-	// we will drop "invalid" data sources (anything other than a PVC or a
-	// VolumeSnapshot) from this request as if an empty PVC had been
-	// requested.
-	if !dataSourceIsEnabled(pvcSpec) {
+// EnforceDataSourceBackwardsCompatibility drops the data source field under certain conditions
+// to maintain backwards compatibility with old behavior.
+// See KEP 1495 for details.
+// Specifically, if this is an update of a PVC with no data source, or a creation of a new PVC,
+// and either the AnyVolumeDataSource feature gate is disabled or the dataSourceRef field is
```
Comment (thockin) [inline pkg/api/persistentvolumeclaim/util.go:44]: Nitty: I wouldn't mention the feature gate here - it's not actually checked and it is irrelevant.  The field is either in use or not.  The gate is handled elsewhere.

Comment (thockin) [reply pkg/api/persistentvolumeclaim/util.go:44]: As discussed in slack - this is the right place to tell the full story about why it is called on both create and update paths, and how it handles back-rev clients

**Review by thockin — COMMENTED** [review]
LGTM from API POV

/approve

**Comment by bswartz** [comment]
I'm going to squash the commits into one so it can merge

**Comment by bswartz** [comment]
Squashed
